lighthouse-core/audits/image-size-responsive.js

```diff
@@ -0,0 +1,198 @@
+    imageRect.bottom >= -ALLOWABLE_OFFSCREEN_Y &&
+    imageRect.left <= viewportDimensions.innerWidth + ALLOWABLE_OFFSCREEN_X &&
+    imageRect.right >= -ALLOWABLE_OFFSCREEN_X
+  );
+}
+
+/**
+ * @param {LH.Artifacts.ImageElement} image
+ * @return {boolean}
+ */
+function isCandidate(image) {
+  if (image.displayedWidth <= 1 || image.displayedHeight <= 1) {
+    return false;
+  }
+  if (image.naturalWidth === 0 || image.naturalHeight === 0) {
+    return false;
+  }
+  if (image.mimeType === 'image/svg+xml') {
+    return false;
+  }
+  if (image.isCss) {
+    return false;
+  }
+  if (image.usesObjectFit) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * @param {LH.Artifacts.ImageElement} image
+ * @param {number} DPR
+ * @return {boolean}
+ */
+function imageHasRightSize(image, DPR) {
+  const [expectedWidth, expectedHeight] =
+      expectedImageSize(image.displayedWidth, image.displayedHeight, DPR);
+  return image.naturalWidth >= expectedWidth && image.naturalHeight >= expectedHeight;
+}
+
+/**
+ * @param {LH.Artifacts.ImageElement} image
+ * @param {number} DPR
+ * @return {Result}
+ */
+function getResult(image, DPR) {
+  const [expectedWidth, expectedHeight] =
+      expectedImageSize(image.displayedWidth, image.displayedHeight, DPR);
+  return {
+    url: image.src,
+    elidedUrl: URL.elideDataURI(image.src),
+    displayedSize: `${image.displayedWidth} x ${image.displayedHeight}`,
+    actualSize: `${image.naturalWidth} x ${image.naturalHeight}`,
+    expectedSize: `${expectedWidth} x ${expectedHeight}`,
+    expectedPixels: expectedWidth * expectedHeight,
+  };
+}
+
+/**
+ * Compute the size an image should have given the display dimensions and pixel density.
+ *
+ * For smaller images, typically icons, the size must be proportional to the density.
+ * For larger images some tolerance is allowed as in thise cases the perceived degradation is not
+ * that bad.
+ *
+ * @param {number} displayedWidth
+ * @param {number} displayedHeight
+ * @param {number} DPR
+ * @return {[number, number]}
+ */
+function expectedImageSize(displayedWidth, displayedHeight, DPR) {
+  let factor = DPR;
+  if (displayedWidth > 64 || displayedHeight > 64) {
+    factor *= TOLERANCE;
+  }
+  const width = Math.ceil(factor * displayedWidth);
+  const height = Math.ceil(factor * displayedHeight);
+  return [width, height];
+}
+
+/**
+ * Remove repeated entries for the same source and sort them by source.
+ *
+ * It will keep the entry with the largest size.
+ *
+ * @param {Result[]} results
+ * @return {Result[]}
+ */
+function deduplicateAndSortResults(results) {
+  results.sort((a, b) => a.url === b.url ? 0 : (a.url < b. url ? -1 : 1));
+  const deduplicated = /** @type {Result[]} */ ([]);
+  for (const r of results) {
+    if (deduplicated.length > 0 && r.url === deduplicated[deduplicated.length - 1].url) {
+      if (deduplicated[deduplicated.length - 1].expectedPixels < r.expectedPixels) {
```

Comment:
```suggestion
      // If the URL was the same, this is a duplicate. Keep the largest image.
      if (deduplicated[deduplicated.length - 1].expectedPixels < r.expectedPixels) {
```

maybe pull out `deduplicated[deduplicated.length - 1]` into `const mostRecentImage` or something for slightly easier groking?